EpsilonAuxBMS/Src/SetContactorsTask.c

```diff
@@ -138,90 +158,123 @@ void setContactorsTask(void const* arg)
                 }
 
-                break;
-            }
-
-            case BLOCKED:
-                if (orionStatus.gpioOk && orionStatus.batteryVoltagesInRange)
+                if (orionStatus.allowCharge && orionStatus.chargeContactorOverriden && !charge)
                 {
-                    state = FIRST_CHECK;
+                    if (osMutexWait(orionStatus.orionStatusMutex, 0) != osOK)
+                    {
+                        continue;
+                    }
+
+                    orionStatus.chargeContactorOverriden = 0;
+                    osMutexRelease(orionStatus.orionStatusMutex);
+
+                    turningOnCharge = 1;
+                    // Turn on charge Contactor and go back to recheck
+                    HAL_GPIO_WritePin(CHARGE_CONTACTOR_ENABLE_GPIO_Port, CHARGE_CONTACTOR_ENABLE_Pin, GPIO_PIN_SET);
+                    state = CHARGE_CONTACTOR_ENABLE_CHECK;
                 }
+                else if (orionStatus.allowDischarge && orionStatus.dischargeContactorOverriden && !discharge)
+                {
+                    if (osMutexWait(orionStatus.orionStatusMutex, 0) != osOK)
+                    {
+                        continue;
+                    }
 
-                break;
+                    orionStatus.dischargeContactorOverriden = 0;
+                    osMutexRelease(orionStatus.orionStatusMutex);
 
-            default:
-                state = FIRST_CHECK;
-        }
-    }
-}
+                    turningOnDischarge = 1;
+                    // Turn on discharge contactor and go back to recheck
+                    HAL_GPIO_WritePin(DISCHARGE_CONTACTOR_ENABLE_GPIO_Port, DISCHARGE_CONTACTOR_ENABLE_Pin, GPIO_PIN_SET);
+                    state = DISCHARGE_CONTACTOR_ENABLE_CHECK;
+                }
 
-int updateContactorState(ContactorState newState, uint8_t error, Contactor contactor)
-{
-    if (osMutexWait(auxStatus.auxStatusMutex, 0) != osOK)
-    {
-        return 0;
-    }
+                if ((orionStatus.allowCharge && !(common && charge) && !turningOnCharge) ||
+                        (orionStatus.allowDischarge && !(common && discharge) && !turningOnDischarge))
+                {
+                    disconnectContactors();
+                    state = CONTACTOR_DISCONNECTED;
+                }
 
-    auxStatus.contactorError = error;
+                break;
+            }
 
-    switch (contactor)
-    {
-        case COMMON:
-            auxStatus.commonContactorState = newState;
-            break;
+            case BLOCKED:
+                common = !HAL_GPIO_ReadPin(COMMON_SENSE_GPIO_Port, COMMON_SENSE_Pin);
+                charge = !HAL_GPIO_ReadPin(CHARGE_SENSE_GPIO_Port, CHARGE_SENSE_Pin);
 
-        case CHARGE:
-            auxStatus.chargeContactorState = newState;
-            break;
+                if ((prevState == DISCHARGE_CONTACTOR_ENABLE_CHECK && !(common && charge)) ||
+                        (prevState == CHARGE_CONTACTOR_ENABLE_CHECK && !common))
```

Comment:
We should be able to simplify this statement to be:
`if ((prevState == DISCHARGE_CONTACTOR_ENABLE_CHECK && !charge) || !common)`

Reply:
We can't, because if the previous state is FIRST_CHECK or COMMON_CONTACTOR_CHECK, we don't want it to shut down and go into CONTACTOR_DISCONNECTED because common hasn't been turned off yet. Any other if statement isolating common gets more confusing since we need more conditions